### PR TITLE
Delete(...) endpoint for the resource service

### PR DESCRIPTION
**Pull request by analogue**
### Description

See NET-2690.

### Testing & Reproduction steps

`gotestsum --format=testname ./agent/grpc-external/services/resource`

### PR Checklist

* [x] updated test coverage
* [ ] external facing docs updated
* [x] not a security concern

### CODE REVIEW

**Review by boxofrad — CHANGES_REQUESTED** [review]
Great work as usual! I've left one minor comment inline, but need to clarify this point:

> Acceptance criteria mentioned having to re-read the resource to get the version, but the version is already passed in via the `DeleteRequest` and seem like punting that off to the storage backend is a better place for that responsibility.

We need to read the resource for cases where `req.Version` is blank — i.e. the majority of user-issued writes on the CLI. In these cases, the user isn't _intentionally_ doing a CAS operation (they just want to delete whatever version of the resource exists) but the storage backend, by design, only supports CAS write and delete operations, partly to keep the interface as small as possible, and to allow for validation hooks in the future.

With this approach, we'll also need to handle `ErrCASFailure` by retrying the whole read-delete loop, but only when the user didn't specify a version. I'm currently implementing this for the `Write` endpoint though, so maybe hold off until we can reuse it here?

**Comment by analogue** [comment]
@boxofrad I merged in `boxofrad/resource-write-endpoint` to get at the demo apis for testing and took another stab at the endpoint. I think there are some gaps in the acceptance criteria so I left a couple of TODOs in the tests. Let's chat when you get a chance.

File: agent/grpc-external/services/resource/delete.go
Comment (boxofrad) [inline agent/grpc-external/services/resource/delete.go:35]: Remember to check for `ErrCASFailure` here too 😉

File: agent/grpc-external/services/resource/delete.go
Comment (boxofrad) [inline agent/grpc-external/services/resource/delete.go:30]: FWIW, there's a chance that the resource will be updated between the `Read` and `DeleteCAS`, in which case the user would see `ErrCASFailure`.

It's pretty unlikely though, so I don't think we have to handle it with retries like we do in the `Write` endpoint. I guess it'd happen if a user was trying to delete a resource at the same time a controller was modifying it.

**Review by boxofrad — COMMENTED** [review]
Nice job 🏆 🎉

**Comment by analogue** [comment]
@boxofrad Ok, changed things up a bit based on a better understanding of the behavior here.

File: agent/grpc-external/services/resource/delete.go
Comment (analogue) [reply agent/grpc-external/services/resource/delete.go]: 👍 Fixed in https://github.com/hashicorp/consul/pull/16756/commits/d038a2538accb28fdad204653df3305e38058fda

File: agent/grpc-external/services/resource/delete.go
Comment (analogue) [reply agent/grpc-external/services/resource/delete.go]: Fixed in https://github.com/hashicorp/consul/pull/16756/commits/d038a2538accb28fdad204653df3305e38058fda

File: agent/grpc-external/services/resource/delete_test.go
Comment (analogue) [reply agent/grpc-external/services/resource/delete_test.go]: 👍 Fixed in https://github.com/hashicorp/consul/pull/16756/commits/d038a2538accb28fdad204653df3305e38058fda

File: agent/grpc-external/services/resource/delete.go
Comment (boxofrad) [inline agent/grpc-external/services/resource/delete.go:47]: I've realized while reading the gRPC status [docs](https://grpc.github.io/grpc/core/md_doc_statuscodes.html) that `Aborted` is a more appropriate code for CAS failures. 
```suggestion
			return nil, status.Error(codes.Aborted, err.Error())
```

File: agent/grpc-external/services/resource/delete.go
Comment (boxofrad) [inline agent/grpc-external/services/resource/delete.go:40]: Sorry to be a pain, but would you mind making the status code explicit here? 🙇🏻‍♂️  (see: https://github.com/hashicorp/consul/pull/16842#discussion_r1157085813)

**Review by boxofrad — APPROVED** [review]
Couple of very small comments, but this is basically good to go (after a rebase). Awesome work! 👏🏻

File: agent/grpc-external/services/resource/delete.go
Comment (analogue) [reply agent/grpc-external/services/resource/delete.go]: 👍 Fixed in https://github.com/hashicorp/consul/pull/16756/commits/24714ca30e26d6ac720f02bdfe4232159f549c2c

File: agent/grpc-external/services/resource/delete.go
Comment (analogue) [reply agent/grpc-external/services/resource/delete.go]: 👍 Fixed in https://github.com/hashicorp/consul/pull/16756/commits/24714ca30e26d6ac720f02bdfe4232159f549c2c

**Pull request merged**
